### ISSUE TITLE
GWT support for LibGDX

### ISSUE THREAD

**Issue by bornskilled200** [ykrasik/jaci#1]
Requesting full support for the gwt for the libgdx backend. I have forked a very hacked version of the source to get it running but if the process is streamlined it would be better for future changes.
- Delombokize the source when uploading to maven
- Replace Java's reflection and instead use LibGDX's wrapper classes.

Delomboking(?) the source would need to be done during the build process.

While it would be very annoying to use LibGDX's wrapper classes in the core, you can instead emulate classes like I lazily did. This will still have some code duplication however.

Also using the latest 1.8.0 for libgdx is preferable because of fixes to it's reflection.

**Comment by ykrasik** [comment]
Hi,

Thanks for your input. I will admit that I haven't tried running Jaci with GWT... or any libGdx project with GWT, so this may take me some time to set up.

Regarding the points you mentioned:
1. Delombok - This shouldn't be an issue, it's possible through Gradle - http://stackoverflow.com/questions/24997441/delombok-using-gradle and if that doesn't work, I'm not opposed to dropping Lombok entirely for the sake of getting this to work.
2. I would imagine that the best way to handle the reflection issues is by having an interface in the core module that will expose an API that both Java & LibGdx reflection support, and access this interface either by passing it to all classes that use it, or by using a provider that will be set to LibGdx variant when using libGdx. I'm leaning towards the provider, even though it's a less 'clean' solution, simply because reflection is also used in static contexts.

In your changes I see that you also replaced all String.format calls to simple string concat, I assume this is because it's unsupported in GWT?

**Comment by bornskilled200** [comment]
I believe you should not need any changes to reflection in the libgdx until GWT backend is being used. This is why I was suggesting "emulation" of certain files. What this actually means in GWT is that during the GWT compilation, it replaces one .java file with another. We could just make ReflectionUtils a bigger part of the project.

You can see this in the jaci-libgdx-cli there is a com.github.ykrasik.emu package which I need to clean out and simplify.

Nice catch on the removal of String.format. GWT does not support String.format and most of regex.

We also need to request for per parameter annotations for the libgdx reflection.

**Comment by ykrasik** [comment]
So far I've removed Lombok, all calls to String.format and regexes from the project.

I'm now trying to handle the last bit with reflection, but the approach I initially wanted to take - "dependency injection", is proving to be a bit cumbersome, as GWT doesn't support anything under java.lang.reflect, requiring me to create my own abstraction over a large portion of the code.

I looked at your emu package, and I see what you did there - It's probably possible to automate the generation of these sources through Gradle, but right now I can only solve this by adding a dependency on libGdx to core, which is a problem.

For now I will continue with the "dependency injection" approach - I only need a small subset of the reflection API, so maybe it won't be so much work eventually.

**Comment by ykrasik** [comment]
0.3.0 was released in order to support this.
- Removed Lombok
- Removed all code not compatible with GWT (String.format, regex)
- Abstracted over reflection code, to allow for a reflection provider (java / libGdx)
- Split jaci-libGdx into 2: jaci-libgdx-cli-java (Java reflection), jaci-libgdx-cli-gwt (libGdx reflection)

Updated the Readme with an explanation.

Thank you for getting this into my attention. Would love to hear if it works.

**Comment by bornskilled200** [comment]
Seems like we LibGdxCli to be in the reflection pool, like so

```
<extend-configuration-property name="gdx.reflect.include" value="com.github.ykrasik.jaci.cli.libgdx.LibGdxCli" />
```

 I'll just submit a pull request cause it's a simple fix.

Otherwise I can't seem to find anything wrong immediately.

**Comment by ykrasik** [comment]
Thanks, however it's working for me when I test it (admittedly, a simple test case). Would you mind sharing the error scenario?

**Comment by bornskilled200** [comment]
That is odd, mine is quite simple also

Test.java

```
package com.unseenspace.test;

import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.utils.viewport.ScreenViewport;
import com.github.ykrasik.jaci.cli.libgdx.LibGdxCli;
import com.github.ykrasik.jaci.cli.libgdx.LibGdxCliBuilder;
import com.github.ykrasik.jaci.cli.libgdx.LibGdxVisibilityToggler;

public class Test extends ApplicationAdapter {
    private Camera camera;

    private ScreenViewport viewport;
    private SpriteBatch batch;

    private Stage stage;

    @Override
    public void create() {
        camera = new OrthographicCamera();

        viewport = new ScreenViewport(camera);
        batch = new SpriteBatch();

        stage = new Stage(viewport, batch);
        Gdx.input.setInputProcessor(stage);

        BasicCommand basicCommand = new BasicCommand();
        LibGdxCli cli = new LibGdxCliBuilder()
                .process(basicCommand)
                .build();
        basicCommand.setCli(cli);
        stage.addActor(cli);
        stage.addListener(new LibGdxVisibilityToggler(cli));
    }

    @Override
    public void resize(int width, int height) {
        viewport.update(width, height, true);
    }

    @Override
    public void render() {
        Gdx.gl.glClearColor(1, 0, 0, 0);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        stage.act();
        stage.draw();
    }
}
```

BasicCommand.java

```
package com.unseenspace.test;

import com.badlogic.gdx.Gdx;
import com.github.ykrasik.jaci.api.Command;
import com.github.ykrasik.jaci.api.CommandOutput;
import com.github.ykrasik.jaci.cli.libgdx.LibGdxCli;

/**
 * Created by madsk_000 on 1/2/2016.
 */
public class BasicCommand {
    private CommandOutput output;
    private LibGdxCli cli;

    public void setCli(LibGdxCli cli) {
        this.cli = cli;
    }

    @Command(description = "Display 'Hello, World!'")
    public void helloWorld() {
        output.message("Hello, World!");
    }

    @Command(description = "Closes the console")
    public void exit() {
        cli.setVisible(false);
    }

    @Command(description = "Quits out of the application")
    public void quit() {
        Gdx.app.exit();
    }
}
```

**Comment by ykrasik** [comment]
Interesting, it happens because BasicCommand contains LibGdxCli as a field.

I didn't want to put all jaci code under GWT Gdx reflection to avoid bloating the cache, I wonder if this error will resurface in the future.

**Comment by ykrasik** [comment]
Haven't updated the README yet, 0.3.1 is out with this fix.

**Comment by bornskilled200** [comment]
I retested by making BasicCommand have a SpriteBatch field which is not in the cache and it causes the same error.

So getting a type for a field, and i'm guessing for any class that is not in the cache, throws an exception of some sort. I feel like this is error in LibGdx though.

I didn't realize that LibGdxCli wasn't needed, it was just a hasty conclusion.

I am still looking at the code and figuring it out. It seems my last pull request was unneeded.

**Comment by ykrasik** [comment]
Actually, this could be a common use case - adding the Cli as a field to a commands class to allow operating on the Cli from within the Cli. Inception? :)

Anyway it's not a very large class so I don't think it's an issue.

I will add to the README that any class referenced as a field in a Commands class, has to be added to the gdx.reflect attribute.

**Comment by bornskilled200** [comment]
Oh yea that would be great, one less class to pass in as a parameter. It does kind of make sense to manipulate the Cli Shell however it could become too specialized in certain backends.

I'm trying to see if there is a bug in libgdx however for now you could just catch an exception in

LibGdxReflectionField

```
    @Override
    public Class<?> getType() { return field.getType(); }
```

and return null instead to sort of hackish fix until the next version of LibGdx.

It would have to be Exception though, unless you want another dependency on GWT

**Comment by ykrasik** [comment]
I see what you're saying, this could become an issue for Commands classes that contain fields required for their own purposes.

I think the less-hackish way to do this would be to change the API of ReflectionField from

```
Class<?> getType();
```

to

```
boolean isType(Class<?>);
```

as this is not a general-purpose reflection API, but one suited to our needs.

It feels less of a hack to say that something isn't of a certain type if an exception is thrown, then to return null.

I'll handle this tomorrow.

**Comment by ykrasik** [comment]
Fixed in 0.3.2

**Comment by bornskilled200** [comment]
Nice!

Everything seems to be working so if there are future problems that we did not foresee, another issue can be made later.

Thank you for this amazing library!
